### PR TITLE
Migrate 'utils/notifications.jsx' and associated tests to TypeScript

### PR DESCRIPTION
**Pull request by karlmarxlopez**
#### Summary

Migrate 'utils/notifications.jsx' and associated tests to TypeScript

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-server/issues/12489

### CODE REVIEW

**Comment by hanzei** [comment]
@karlmarxlopez Same here

**Comment by karlmarxlopez** [comment]
@hanzei Hi, do I need to ignore `mattermost-redux`s types until the `typescript` migration is ready? Just to pass the `type-check` stage.

**Comment by karlmarxlopez** [comment]
@hanzei This is ready for review

**Comment by devinbinnie** [comment]
Looks like a few lint errors are still present, can you clear those up?
```
/root/mattermost-webapp/components/edit_post_modal/edit_post_modal.jsx
  9:1   error  There should be at least one empty line between import groups  import/order
  9:43  error  Unable to resolve path to module 'utils/constants.jsx'         import/no-unresolved

/root/mattermost-webapp/utils/utils.jsx
   400:5   error  The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype  guard-for-in
  2076:12  error  Unexpected negated condition                                                                                no-negated-condition
  2179:47  error  Unexpected mix of '+' and '*'                                                                               no-mixed-operators
  2179:51  error  Unexpected mix of '+' and '*'                                                                               no-mixed-operators
```

**Comment by karlmarxlopez** [comment]
> Looks like a few lint errors are still present, can you clear those up?
> 
> ```
> /root/mattermost-webapp/components/edit_post_modal/edit_post_modal.jsx
>   9:1   error  There should be at least one empty line between import groups  import/order
>   9:43  error  Unable to resolve path to module 'utils/constants.jsx'         import/no-unresolved
> 
> /root/mattermost-webapp/utils/utils.jsx
>    400:5   error  The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype  guard-for-in
>   2076:12  error  Unexpected negated condition                                                                                no-negated-condition
>   2179:47  error  Unexpected mix of '+' and '*'                                                                               no-mixed-operators
>   2179:51  error  Unexpected mix of '+' and '*'                                                                               no-mixed-operators
> ```

@devinbinnie These lines are disabled, I don't know why `eslint-disable-line` is not working?

**Comment by karlmarxlopez** [comment]
@devinbinnie Hi, I have only 1 failing stage.

**Comment by karlmarxlopez** [comment]
@hanzei Hi, I have only 1 failing stage.

File: components/login/login_controller/login_controller.jsx [hunk 1]
```diff
@@ -62,14 +62,20 @@ class LoginController extends React.Component {
...
-    }
+    };
 
     constructor(props) {
         super(props);
 
         let loginId = '';
-        if ((new URLSearchParams(this.props.location.search)).get('extra') === Constants.SIGNIN_VERIFIED && (new URLSearchParams(this.props.location.search)).get('email')) {
-            loginId = (new URLSearchParams(this.props.location.search)).get('email');
+        if (
```
Comment (jespino) [inline components/login/login_controller/login_controller.jsx]: You are reformating a lot of code to and adding a lot of line breaks which is inconsistent with the current code styling. I would prefer to keep untouch the code that is not related to the PR in that sense, actually the code should be consistent with the rest fo the code base, and this "80 characters" approach is breaking the current existing style.

Comment (karlmarxlopez) [reply components/login/login_controller/login_controller.jsx]: Sorry, I've misconfigured my VSCode eslint settings that auto-formats the code and ignores the `.eslintrc`. I'll fix this first thing tomorrow.

File: components/login/login_controller/login_controller.jsx [hunk 2]
```diff
@@ -8,15 +8,15 @@ import {Link} from 'react-router-dom';
```
Comment (jespino) [inline components/login/login_controller/login_controller.jsx]: A lot of formatting changes in this file in this PR not related to the changes expected from the PR. A lot of styling changes in the code.

File: components/system_notice/index.ts [hunk 3]
```diff
@@ -3,16 +3,28 @@
```
Comment (jespino) [inline components/system_notice/index.ts]: Why this file is changed to typescript while is not related to the PR?

File: types/key-mirror.d.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+
```
Comment (jespino) [inline types/key-mirror.d.ts]: I like this :)

File: utils/__snapshots__/message_html_to_component.test.jsx.snap [hunk 5]
```diff
@@ -45,10 +45,12 @@ Array [
   "
```
Comment (jespino) [inline utils/__snapshots__/message_html_to_component.test.jsx.snap]: Changes in this file probably related to not needed reformating.

File: utils/utils.jsx [hunk 6]
```diff
@@ -8,27 +8,38 @@ import semver from 'semver';
```
Comment (jespino) [inline utils/utils.jsx]: A lot of not related style reformating in this file.

File: utils/__snapshots__/message_html_to_component.test.jsx.snap (hunk 5 above)
Comment (karlmarxlopez) [reply utils/__snapshots__/message_html_to_component.test.jsx.snap]: I did not change anything on `message_html_to_component`, I've just pulled the upstream master into this branch. And ran, `npm test`

File: package.json [hunk 7]
```diff
@@ -96,6 +96,7 @@
     "@types/react-bootstrap": "^0.32.20",
     "@types/react-color": "3.0.1",
     "@types/react-intl": "2.3.18",
+    "@types/react-redux": "^7.1.4",
```
Comment (jespino) [inline package.json]: We use specific versions, please remove the `^` here.

File: utils/notifications.test.jsx [hunk 8]
```diff
@@ -8,31 +8,39 @@ describe('Notifications.showNotification', () => {
```
Comment (jespino) [inline utils/notifications.test.jsx]: Not needed formatting changes in this file, not a big deal to keep the changes

File: utils/notifications.tsx [hunk 9]
```diff
@@ -28,15 +50,18 @@ export async function showNotification({title, body, requireInteraction, silent,
         throw new Error('Notification.requestPermission not supported');
     }
 
-    if (Notification.permission !== 'granted' && requestedNotificationPermission) {
+    if (
```
Comment (jespino) [inline utils/notifications.tsx]: More not needed format changes

File: utils/notifications.tsx [hunk 10]
```diff
@@ -28,15 +50,18 @@ export async function showNotification({title, body, requireInteraction, silent,
...
+    requestedNotificationPermission
+    ) {
         throw new Error('Notifications already requested but not granted');
     }
 
     requestedNotificationPermission = true;
 
     let permission = await Notification.requestPermission();
     if (typeof permission === 'undefined') {
-        // Handle browsers that don't support the promise-based syntax.
```
Comment (jespino) [inline utils/notifications.tsx]: More not needed format changes

File: utils/notifications.tsx [hunk 11]
```diff
@@ -14,7 +14,29 @@ let requestedNotificationPermission = false;
 // notification. Notifications that do not require interaction will be closed automatically after
 // the Constants.DEFAULT_NOTIFICATION_DURATION. Not all platforms support all features, and may
 // choose different semantics for the notifications.
-export async function showNotification({title, body, requireInteraction, silent, onClick} = {}) {
+
+export interface ShowNotificationParams {
```
Comment (jespino) [inline utils/notifications.tsx]: We are using type aliases for this cases, probably something more consistent would be `type Props = {...`, besides you are exporting but there is no need to export it.

Comment (jespino) [reply utils/notifications.tsx]: Nevermind, this is correct, because is not a component, is a function.

**Comment by karlmarxlopez** [comment]
Hi @jespino sure, thanks also for being patient with me. I'm just a n00b on open-source community.

**Comment by hanzei** [comment]
Hey @karlmarxlopez 

You have one linter error:
```
/root/mattermost-webapp/actions/global_actions.jsx
  42:1  error  'utils/constants' import is duplicated  no-duplicate-imports

✖ 1 problem (1 error, 0 warnings)
```

File: actions/global_actions.jsx [hunk 12]
```diff
@@ -36,10 +36,9 @@ import store from 'stores/redux_store.jsx';
 import LocalStorageStore from 'stores/local_storage_store';
 import WebSocketClient from 'client/web_websocket_client.jsx';
 
-import {ActionTypes, Constants, PostTypes, RHSStates} from 'utils/constants.jsx';
+import {ActionTypes, Constants, PostTypes, RHSStates, ModalIdentifiers} from 'utils/constants';
 import {filterAndSortTeamsByDisplayName} from 'utils/team_utils.jsx';
 import * as Utils from 'utils/utils.jsx';
-import {ModalIdentifiers} from 'utils/constants';
```
Comment (saturninoabril) [inline actions/global_actions.jsx]: Good catch!

File: actions/websocket_actions.test.jsx [hunk 13]
```diff
@@ -24,7 +23,7 @@ import store from 'stores/redux_store.jsx';
 import configureStore from 'tests/test_store';
 
 import {browserHistory} from 'utils/browser_history';
-import Constants, {SocketEvents, UserStatuses} from 'utils/constants';
+import Constants, {SocketEvents, UserStatuses, ActionTypes} from 'utils/constants';
```
Comment (saturninoabril) [inline actions/websocket_actions.test.jsx]: 👍
